### PR TITLE
prysm scripts should not save 404 pages

### LINKED ISSUE #9041: prysm.sh, prysm.bat, prysm.ps1 should not save 404 pages
# 🐞 Bug Report

### Description

When the prysm download scripts access a 404 page. The script should not save the 404 page and try to continue through the script. The script should fail immediately when hitting a 404.

### Has this worked before in a previous version?

N/A

## 🔬 Minimal Reproduction

USE_PRYSM_VERSION=fake ./prysm.sh beacon-chain version

## 🔥 Error

<pre><code>Latest Prysm version is fake.
Downloading beacon chain@fake to /home/preston/git/prysm/dist/beacon-chain-fake-linux-amd64 (specified in $USE_PRYSM_VERSION)
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100  2337    0  2337    0     0  10869      0 --:--:-- --:--:-- --:--:-- 10869
Verifying binary integrity.
shasum: /home/preston/git/prysm/dist/beacon-chain-fake-linux-amd64.sha256: no properly formatted SHA checksum lines found
Failed to verify Prysm binary. Please erase downloads in the
dist directory and run this script again. Alternatively, you can use a
A prior version by specifying environment variable USE_PRYSM_VERSION
with the specific version, as desired. Example: USE_PRYSM_VERSION=v1.0.0-alpha.5
If you must wish to continue running an unverified binary, specific the
environment variable PRYSM_ALLOW_UNVERIFIED_BINARIES=1
</code></pre>

## 🌍  Your Environment

**Operating System:**

<pre>
  <code>
N/A
  </code>
</pre>

**What version of Prysm are you running? (Which release)**

<pre>
  <code>
N/A
  </code>
</pre>

**Anything else relevant (validator index / public key)?**

### PR DESCRIPTION
**Pull request by ahadda5**
**What type of PR is this?**

Bug Fix

**What does this PR do? Why is it needed?**
prysm.sh, .bat and .ps1 should not save 404 pages . Currently the curl command does not check if a 404 is returned and continues with the sha256 and signature files for verification. It is on verification which it then fails. 
A check was added to exit if a 404 error is encountered. 

**Which issues(s) does this PR fix?**

Part of #9041

**Testing**
**Linux** test case (ensure dist folder does not contain the file, e.g. validator-fake-linux-amd64,  which curl is trying to retrieve) .After this change is implemented no longer will it save a 404 file. 
This changes tackles curl 404 error only. 
```
~/prysm$ USE_PRYSM_VERSION=fake ./prysm.sh validator version
Latest Prysm version is fake.
Downloading validator@fake to /home/ahaddad/prysm/dist/validator-fake-linux-amd64 (specified in $USE_PRYSM_VERSION)
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100  2337    0  2337    0     0   4015      0 --:--:-- --:--:-- --:--:--  4008
No prysm validator found for fake, (validator-fake-linux-amd64) exit
```

```
>USE_PRYSM_VERSION=fake ./prysm.sh client-stats version
>USE_PRYSM_VERSION=fake ./prysm.sh beacon-chain version
```
Try the same with a true version or let the .sh determine it. 
`> ./prysm.sh beacon-chain version`

**Windows**  .bat file .  Unstable .bat on my machine.  

Ensure that the dist folder does not contain an older fake file. 

1. open a cmd (either admin or regular )
2. under the prysm folder run this. Not running this extra code it will cause the actual script to complain `( was unexpected at this time.`  I think this is a glitch with curl/bat on windows. I'll check with Microsoft.  [need to resolve this!]
`>for /f "delims=" %i in ('curl --silent -w "%{http_code}" -L https://prysmaticlabs.com/releases/beacon-chain-%prysm_version%-%system%-%arch% -o gg.txt') do set http=%i`
3. >set USE_PRYSM_VERSION=fake
4. > prysm.bat beacon-chain version

The cmd screen will close(exit) after finding a 400. 
The same test will be repeated with > set USE_PRYSM_VERSION=v1.3.11  where it will continue and ask to 'press any key to continue'

_Note that  GoTos where added at the end of every if statement checking the first arg $1 because it was going through each of them._ Very weird. (that is the existing script before the change).

**Windows** .ps1 file
It works as expected in .ps1 . No changes required. 

```
> .\prysm.ps1 beacon-chain version


    Directory: C:\prysm


Mode                 LastWriteTime         Length Name
----                 -------------         ------ ----
d-----         6/26/2021   3:25 PM                dist
Detected variable $env:USE_PRYSM_VERSION=fake
Downloading beacon-chain-fake-windows-amd64.exe
ERROR: Failed to get download beacon-chain-fake-windows-amd64.exe:
  The remote server returned an error: (404) Not Found.
```

**Other notes for review**

### CODE REVIEW

**Comment by ahadda5** [comment]
still .bat and .ps1 need to be added.

File: prysm.sh [hunk 1]
```diff
@@ -181,7 +181,11 @@ if [[ $1 == beacon-chain ]]; then
     if [[ ! -x $BEACON_CHAIN_REAL ]]; then
         color "34" "Downloading beacon chain@${prysm_version} to ${BEACON_CHAIN_REAL} (${reason})"
         file=beacon-chain-${prysm_version}-${system}-${arch}
-        curl -L "https://prysmaticlabs.com/releases/${file}" -o "$BEACON_CHAIN_REAL"
+        res=$(curl -w '%{http_code}\n' -L "https://prysmaticlabs.com/releases/${file}"  -o "$BEACON_CHAIN_REAL")
+        if [[ $res == 404 ]];then
```
Comment (terencechain) [inline prysm.sh:185]: Define 404 as a variable and compare it through out?

File: prysm.sh [hunk 2]
```diff
@@ -181,7 +181,12 @@ if [[ $1 == beacon-chain ]]; then
...
+          echo "404 error, exit"
```
Comment (prestonvanloon) [inline prysm.sh:186]: Can you please add a more meaningful message? For example:
```
No prysm beacon chain found for release $prysm_version
```

**Comment by prestonvanloon** [comment]
Another note: I tested this locally and it still saved the 404 page.

**Comment by ahadda5** [comment]
> Another note: I tested this locally and it still saved the 404 page.

Yes it runs curl first, then based on the http result we decide whether to continue or not.   So you want to first curl -I (the head info) and if that is good then curl to download?

File: prysm.sh (hunk 2 above)
Comment (ahadda5) [reply prysm.sh]: addressed.

**Comment by ahadda5** [comment]
> Another note: I tested this locally and it still saved the 404 page.

addressed.  Needed to use -f option (to fail on non 2xx/3xx errors) and grep.

**Comment by ahadda5** [comment]
I pushed it for review.  The .bat is weird.  One option is to unify the startup script to be in go and just wrap call it from .sh/.bat/.ps1 .

**Review by prestonvanloon — APPROVED** [review]
Looks good to me! Tested locally and I did not observe any files saved in my dist directory when providing an invalid version

**Comment by ahadda5** [comment]
Yes .Sh works as intended.  It's the . Bat that is weird. I describe the behavior above even before a change (and I have the latest windows)

. Ps1 requires no change.

**Pull request merged**